### DISCUSSION TITLE
Zig Style comp time code

### DISCUSSION THREAD

**Discussion by bryanerayner** [dotnet/csharplang#7427] [Language Ideas]
Right now, C# generators are the only way to add efficient AoT code to enable powerful DSLs for things like Dapper, Json encode/ decode, and the like. As far as I know, they're the closest thing we have to macros for C#.

In Zig, the comptime feature allows you to write macros and custom code which is added to the zig compilation - in Zig.

While we have this capability afforded to us by the generator system, it is not intuitive. Can we add a language keyword which would act as sugar around generators, and make them as ubiquitous to use as generics are today?

**Comment by HaloFour** [comment]
See: #1028

**Reply by HaloFour** [reply]
I also believe this case is covered by a combination of interceptors and generators, the former of which is going to be released in preview for C# 12.

**Reply by bryanerayner** [reply]
This is not in the same vein though it could be used for trivial things like figuring out the length of a string constant.

I'm more interested in being able to use the reflection libraries at compile time and emit C# code which is included in the final compilation. Again, I don't think this needs to be introduce new concepts, we have the concepts already. What I'm after, is broadening the ease of access so that more developers can use these features.
